# 评估和清理TED数据

## 分析目标

此数据分析的目的是，根据TED数据，挖掘最受欢迎的演讲主题，以便分析当今社会人群的关注热点。

本实战项目的目的在于练习评估数据干净和整洁度，并且基于评估结果，对数据进行清洗，从而得到可供下一步分析的数据。

## 简介

该数据集是国际思想传播平台 TED（Ideas Worth Spreading）演讲的结构化汇总数据，系统记录了多场 TED 演讲的基础信息、演讲者背景、传播表现及受众反馈，是研究全球思想传播趋势、公众兴趣偏好及 TED 平台生态的优质样本。

1. 演讲基础信息
- `title`：演讲标题，直观概括演讲核心议题；
- `name`：演讲的完整名称（含演讲者与主题），便于快速定位演讲身份。
- `description`：演讲内容简介，详细说明演讲者观点或演讲主旨；
- `duration`：演讲时长（单位：秒）；
- `event`：演讲所属 TED 活动，标记演讲的举办场景与年份；
- `film_date/published_date`：均为 Unix 时间戳（秒级），分别对应演讲拍摄日期与平台发布日期；
2. 演讲者信息
- `main_speaker`：主要演讲者姓名，涵盖科学家、社会活动家、艺术家、企业家等多元身份；
- `speaker_occupation`：演讲者职业，反映演讲者的专业背景；
- `num_speaker`：单场演讲的演讲者人数（默认值为 1，少数为多人合作演讲）。
3. 传播与互动表现
- `views`：演讲观看次数，范围从数万到数千万，直接体现演讲的传播影响力；
- `comments`：演讲评论数量，反映受众对演讲的讨论热度；
- `languages`：演讲提供的翻译语言种类，体现演讲的全球传播广度。
4. 受众反馈与主题分类
- `ratings`：受众评价列表，包含 14 类评价维度及对应评价次数，可用于分析受众偏好；
- `tags`：演讲主题标签，涵盖教育、科技、环境、心理学、文化等领域，是主题挖掘的核心字段；
- `related_talks`：相关演讲列表，记录与当前演讲主题关联的其他演讲信息（如演讲 ID、标题、观看量），体现内容间的关联性；
- `url`：演讲在 TED 官网的直接链接，支持跳转查看原始演讲视频或页面。

## 读取数据

导入数据分析所需要的库，利用pandas的`read_csv`函数，将原始数据文件“ted.csv”的数据内容导入，并解析为DataFrame，并将其赋值给`ted_data`

In [31]:
import pandas as pd

In [32]:
ted_data = pd.read_csv("ted.csv")
ted_data.head()

,comments,description,duration,event,film_date,languages,main_speaker,name,num_speaker,published_date,ratings,related_talks,speaker_occupation,tags,title,url,views
0,4553,Sir Ken Robinson makes an entertaining and profoundly moving case for creating an education system that nurtures (rather than undermines) creativity.,1164,TED2006,1140825600,60,Ken Robinson,Ken Robinson: Do schools kill creativity?,1,1151367060,"[{'id': 7, 'name': 'Funny', 'count': 19645}, {'id': 1, 'name': 'Beautiful', 'count': 4573}, {'id': 9, 'name': 'Ingenious', 'count': 6073}, {'id': 3, 'name': 'Courageous', 'count': 3253}, {'id': 11, 'name': 'Longwinded', 'count': 387}, {'id': 2, 'name': 'Confusing', 'count': 242}, {'id': 8, 'name...","[{'id': 865, 'hero': 'https://pe.tedcdn.com/images/ted/172559_800x600.jpg', 'speaker': 'Ken Robinson', 'title': 'Bring on the learning revolution!', 'duration': 1008, 'slug': 'sir_ken_robinson_bring_on_the_revolution', 'viewed_count': 7266103}, {'id': 1738, 'hero': 'https://pe.tedcdn.com/images/...",Author/educator,"['children', 'creativity', 'culture', 'dance', 'education', 'parenting', 'teaching']",Do schools kill creativity?,https://www.ted.com/talks/ken_robinson_says_schools_kill_creativity\n,47227110
1,265,"With the same humor and humanity he exuded in ""An Inconvenient Truth,"" Al Gore spells out 15 ways that individuals can address climate change immediately, from buying a hybrid to inventing a new, hotter brand name for global warming.",977,TED2006,1140825600,43,Al Gore,Al Gore: Averting the climate crisis,1,1151367060,"[{'id': 7, 'name': 'Funny', 'count': 544}, {'id': 3, 'name': 'Courageous', 'count': 139}, {'id': 2, 'name': 'Confusing', 'count': 62}, {'id': 1, 'name': 'Beautiful', 'count': 58}, {'id': 21, 'name': 'Unconvincing', 'count': 258}, {'id': 11, 'name': 'Longwinded', 'count': 113}, {'id': 8, 'name': ...","[{'id': 243, 'hero': 'https://pe.tedcdn.com/images/ted/566c14767bd62c5ff760e483c5b16cd2753328cd_2880x1620.jpg', 'speaker': 'Al Gore', 'title': 'New thinking on the climate crisis', 'duration': 1674, 'slug': 'al_gore_s_new_thinking_on_the_climate_crisis', 'viewed_count': 1751408}, {'id': 547, 'he...",Climate advocate,"['alternative energy', 'cars', 'climate change', 'culture', 'environment', 'global issues', 'science', 'sustainability', 'technology']",Averting the climate crisis,https://www.ted.com/talks/al_gore_on_averting_climate_crisis\n,3200520
2,124,"New York Times columnist David Pogue takes aim at technology’s worst interface-design offenders, and provides encouraging examples of products that get it right. To funny things up, he bursts into song.",1286,TED2006,1140739200,26,David Pogue,David Pogue: Simplicity sells,1,1151367060,"[{'id': 7, 'name': 'Funny', 'count': 964}, {'id': 3, 'name': 'Courageous', 'count': 45}, {'id': 9, 'name': 'Ingenious', 'count': 183}, {'id': 1, 'name': 'Beautiful', 'count': 60}, {'id': 21, 'name': 'Unconvincing', 'count': 104}, {'id': 11, 'name': 'Longwinded', 'count': 78}, {'id': 8, 'name': '...","[{'id': 1725, 'hero': 'https://pe.tedcdn.com/images/ted/b7f415a054cc0a2bfdd90d0ad5a7f64cf060150d_1600x1200.jpg', 'speaker': 'David Pogue', 'title': '10 top time-saving tech tips', 'duration': 344, 'slug': 'david_pogue_10_top_time_saving_tech_tips', 'viewed_count': 4843421}, {'id': 2274, 'hero': ...",Technology columnist,"['computers', 'entertainment', 'interface design', 'media', 'music', 'performance', 'simplicity', 'software', 'technology']",Simplicity sells,https://www.ted.com/talks/david_pogue_says_simplicity_sells\n,1636292
3,200,"In an emotionally charged talk, MacArthur-winning activist Majora Carter details her fight for environmental justice in the South Bronx -- and shows how minority neighborhoods suffer most from flawed urban policy.",1116,TED2006,1140912000,35,Majora Carter,Majora Carter: Greening the ghetto,1,1151367060,"[{'id': 3, 'name': 'Courageous', 'count': 760}, {'id': 1, 'name': 'Beautiful', 'count': 291}, {'id': 2, 'name': 'Confusing', 'count': 32}, {'i

## 评估数据

这一步，我将对上部建立的`ted_data`这个DataFrame所包含的数据进行评价。

评估主要从两个方面进行：结构和内容，即整齐度和干净度。数据的结构性问题指的是不符合“每列是一个变量，每行是一个观察值，每个单元格是一个值”这三个标准，数据的内容性问题是指存在丢失数据、重复数据、无效数据等。

### 评估数据的整齐度

In [33]:
ted_data.sample(5)

,comments,description,duration,event,film_date,languages,main_speaker,name,num_speaker,published_date,ratings,related_talks,speaker_occupation,tags,title,url,views
177,40,"Paola Antonelli, design curator at New York's Museum of Modern Art, wants to spread her appreciation of design -- in all shapes and forms -- around the world.",1097,TED2007,1172880000,22,Paola Antonelli,Paola Antonelli: Treat design as art,1,1200367200,"[{'id': 24, 'name': 'Persuasive', 'count': 41}, {'id': 8, 'name': 'Informative', 'count': 138}, {'id': 25, 'name': 'OK', 'count': 83}, {'id': 2, 'name': 'Confusing', 'count': 27}, {'id': 11, 'name': 'Longwinded', 'count': 68}, {'id': 10, 'name': 'Inspiring', 'count': 111}, {'id': 3, 'name': 'Cou...","[{'id': 372, 'hero': 'https://pe.tedcdn.com/images/ted/56215_480x360.jpg', 'speaker': 'Paola Antonelli', 'title': 'Design and the Elastic Mind', 'duration': 1060, 'slug': 'paola_antonelli_previews_design_and_the_elastic_mind', 'viewed_count': 313572}, {'id': 266, 'hero': 'https://pe.tedcdn.com/i...",Design curator,"['art', 'culture', 'design', 'museums']",Treat design as art,https://www.ted.com/talks/paola_antonelli_treats_design_as_art\n,585052
1601,124,"How do we solve the problem of the suburbs? Urbanist Jeff Speck shows how we can free ourselves from dependence on the car -- which he calls ""a gas-belching, time-wasting, life-threatening prosthetic device"" -- by making our cities more walkable and more pleasant for more people.",1016,TEDCity2.0,1379635200,26,Jeff Speck,Jeff Speck: The walkable city,1,1381763957,"[{'id': 26, 'name': 'Obnoxious', 'count': 10}, {'id': 21, 'name': 'Unconvincing', 'count': 19}, {'id': 25, 'name': 'OK', 'count': 41}, {'id': 11, 'name': 'Longwinded', 'count': 12}, {'id': 7, 'name': 'Funny', 'count': 55}, {'id': 22, 'name': 'Fascinating', 'count': 92}, {'id': 10, 'name': 'Inspi...","[{'id': 1838, 'hero': 'https://pe.tedcdn.com/images/ted/f5cdecc9ae7a5f2472047e17c233d68e7d172264_1600x1200.jpg', 'speaker': 'Janette Sadik-Khan', 'title': ""New York's streets? Not so mean any more"", 'duration': 842, 'slug': 'janette_sadik_khan_new_york_s_streets_not_so_mean_any_more', 'viewed_co...",Urban planner,"['cities', 'culture', 'design', 'transportation']",The walkable city,https://www.ted.com/talks/jeff_speck_the_walkable_city\n,1041732
858,58,"Diagnosed with cancer, Bruce Feiler worried first about his young family. So -- as he shares in this funny, rambling and ultimately thoughtful talk -- he asked his closest friends to become a ""council of dads,"" bringing their own lifetimes of wisdom to advise his twin daughters as they grow.",1233,TEDMED 2010,1286668800,20,Bruce Feiler,Bruce Feiler: The council of dads,1,1296149280,"[{'id': 10, 'name': 'Inspiring', 'count': 278}, {'id': 1, 'name': 'Beautiful', 'count': 112}, {'id': 3, 'name': 'Courageous', 'count': 114}, {'id': 11, 'name': 'Longwinded', 'count': 38}, {'id': 7, 'name': 'Funny', 'count': 26}, {'id': 25, 'name': 'OK', 'count': 32}, {'id': 22, 'name': 'Fascinat...","[{'id': 769, 'hero': 'https://pe.tedcdn.com/images/ted/7588adb5a757c04f01d78b579f2b6d941d3229ba_2880x1620.jpg', 'speaker': 'Aimee Mullins', 'title': 'The opportunity of adversity', 'duration': 1318, 'slug': 'aimee_mullins_the_opportunity_of_adversity', 'viewed_count': 2054178}, {'id': 1048, 'her...",Writer,"['cancer', 'community', 'culture', 'storytelling']",The council of dads,https://www.ted.com/talks/bruce_feiler_the_council_of_dads\n,465856
17,88,Fumes from indoor cooking fires kill more than 2 million children a year in the developing world. MIT engineer Amy Smith details an exciting but simple solution: a tool for turning farm waste into clean-burning charcoal.,906,TED2006,1140739200,27,Amy Smith,Amy Smith: Simple designs to save a life,1,1155600660,"[{'id': 9, 'name': 'Ingenious', 'count': 321}, {'id': 3, 'name': 'Courageous', 'count': 150}, {'id': 1, 'name': 'Beautiful', 'count': 67}, {'id': 8, 'name': 'Informative', 'count': 279}, {'id': 10, 'name': 'Inspiring', 'coun

由于部分列显示不全，可以适当调整列宽。

In [34]:
pd.set_option("display.max_colwidth", 300)
ted_data.sample(5)

,comments,description,duration,event,film_date,languages,main_speaker,name,num_speaker,published_date,ratings,related_talks,speaker_occupation,tags,title,url,views
1801,293,"""We spend so much time listening to the things people are saying that we rarely pay attention to the things they don't,"" says poet and teacher Clint Smith. A short, powerful piece from the heart, about finding the courage to speak up against ignorance and injustice.",258,TED@NYC,1404777600,40,Clint Smith,Clint Smith: The danger of silence,1,1408117565,"[{'id': 1, 'name': 'Beautiful', 'count': 1544}, {'id': 3, 'name': 'Courageous', 'count': 1355}, {'id': 10, 'name': 'Inspiring', 'count': 3182}, {'id': 22, 'name': 'Fascinating', 'count': 367}, {'id': 2, 'name': 'Confusing', 'count': 45}, {'id': 24, 'name': 'Persuasive', 'count': 592}, {'id': 8, ...","[{'id': 1625, 'hero': 'https://pe.tedcdn.com/images/ted/77f51be04c3531bb9b22eb93029eb01a3d991ff1_1600x1200.jpg', 'speaker': 'Lemon Andersen', 'title': ""Please don't take my Air Jordans"", 'duration': 535, 'slug': 'lemon_andersen_performs_please_don_t_take_my_air_jordans', 'viewed_count': 1120289}...","Poet, educator","['poetry', 'social change', 'spoken word']",The danger of silence,https://www.ted.com/talks/clint_smith_the_danger_of_silence\n,3804535
2164,47,"What if technology could connect us more deeply with our surroundings instead of distracting us from the real world? With the Meta 2, an augmented reality headset that makes it possible for users to see, grab and move holograms just like physical objects, Meron Gribetz hopes to extend our senses...",654,TED2016,1455667200,25,Meron Gribetz,Meron Gribetz: A glimpse of the future through an augmented reality headset,1,1458312270,"[{'id': 21, 'name': 'Unconvincing', 'count': 57}, {'id': 8, 'name': 'Informative', 'count': 174}, {'id': 22, 'name': 'Fascinating', 'count': 288}, {'id': 10, 'name': 'Inspiring', 'count': 185}, {'id': 9, 'name': 'Ingenious', 'count': 134}, {'id': 23, 'name': 'Jaw-dropping', 'count': 108}, {'id':...","[{'id': 38, 'hero': 'https://pe.tedcdn.com/images/ted/e87f6855a745a988c2dee9ceba6bfbd4f854a3a6_1600x1200.jpg', 'speaker': 'Ray Kurzweil', 'title': 'The accelerating power of technology', 'duration': 1376, 'slug': 'ray_kurzweil_on_how_technology_will_transform_us', 'viewed_count': 2434473}, {'id'...","Founder and CEO, Meta","['Senses', 'augmented reality', 'brain', 'computers', 'creativity', 'cyborg', 'demo', 'design', 'engineering', 'entrepreneur', 'innovation', 'interface design', 'invention', 'neuroscience', 'potential', 'prediction', 'product design', 'technology', 'visualizations']",A glimpse of the future through an augmented reality headset,https://www.ted.com/talks/meron_gribetz_a_glimpse_of_the_future_through_an_augmented_reality_headset\n,1575381
265,137,"Chris Abani tells stories of people: People standing up to soldiers. People being compassionate. People being human and reclaiming their humanity. It's ""ubuntu,"" he says: the only way for me to be human is for you to reflect my humanity back at me.",974,TED2008,1201910400,28,Chris Abani,Chris Abani: On humanity,1,1216688400,"[{'id': 3, 'name': 'Courageous', 'count': 491}, {'id': 1, 'name': 'Beautiful', 'count': 591}, {'id': 23, 'name': 'Jaw-dropping', 'count': 135}, {'id': 10, 'name': 'Inspiring', 'count': 672}, {'id': 7, 'name': 'Funny', 'count': 127}, {'id': 22, 'name': 'Fascinating', 'count': 126}, {'id': 8, 'nam...","[{'id': 155, 'hero': 'https://pe.tedcdn.com/images/ted/40a5d6a5fd936376fdd9ce987b73f4fb744b2ea7_2880x1620.jpg', 'speaker': 'Chris Abani', 'title': 'Telling stories from Africa', 'duration': 1056, 'slug': 'chris_abani_on_the_stories_of_africa', 'viewed_count': 561705}, {'id': 64, 'hero': 'https:/...","Novelist, poet","['Africa', 'culture', 'entertainment', 'humanity', 'humor', 'literature', 'poetry', 'writing']",On humanity,https://www.ted.com/talks/chris_abani_muses_on_humanity\n,843859
853,114,"In this first-of-its-kind demo, Heather Knight introduces Dat

通过随机的5行观察值进行分析，符合“每列是一个变量，每列是一个观察值”，但是不符合“每个单元格是一个值”，其中`ratings`、`related_talks`、`speaker_occupation`、`tags`四列对应各行观察值均包含多个变量，后续数据清理时需拆分为单一变量形式。其余无需调整。

### 评估数据的干净度

In [35]:
ted_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2550 entries, 0 to 2549
Data columns (total 17 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   comments            2550 non-null   int64 
 1   description         2550 non-null   object
 2   duration            2550 non-null   int64 
 3   event               2550 non-null   object
 4   film_date           2550 non-null   int64 
 5   languages           2550 non-null   int64 
 6   main_speaker        2550 non-null   object
 7   name                2550 non-null   object
 8   num_speaker         2550 non-null   int64 
 9   published_date      2550 non-null   int64 
 10  ratings             2550 non-null   object
 11  related_talks       2550 non-null   object
 12  speaker_occupation  2544 non-null   object
 13  tags                2550 non-null   object
 14  title               2550 non-null   object
 15  url                 2550 non-null   object
 16  views               2550

从输出结果看，共有2550条观察值，其中`speaker_occupation`变量存在缺失值。

另外，`film_date`和`published_date`的数据类型应为日期，应当进行数据格式转换。

#### 评估缺失数据

了解`speaker_occupation`存在缺失值后，根据条件筛选出缺失值。

In [36]:
ted_data[ted_data["speaker_occupation"].isnull()]

,comments,description,duration,event,film_date,languages,main_speaker,name,num_speaker,published_date,ratings,related_talks,speaker_occupation,tags,title,url,views
1113,145,"After a crisis, how can we tell if water is safe to drink? Current tests are slow and complex, and the delay can be deadly, as in the cholera outbreak after Haiti's earthquake in 2010. TED Fellow Sonaar Luthra previews his design for a simple tool that quickly tests water for safety -- the Water...",217,TEDGlobal 2011,1310601600,38,Sonaar Luthra,Sonaar Luthra: Meet the Water Canary,1,1326731605,"[{'id': 10, 'name': 'Inspiring', 'count': 73}, {'id': 8, 'name': 'Informative', 'count': 85}, {'id': 9, 'name': 'Ingenious', 'count': 167}, {'id': 23, 'name': 'Jaw-dropping', 'count': 27}, {'id': 21, 'name': 'Unconvincing', 'count': 24}, {'id': 24, 'name': 'Persuasive', 'count': 36}, {'id': 25, ...","[{'id': 523, 'hero': 'https://pe.tedcdn.com/images/ted/d9441da0244acb76f766b462f1674d4310cc6fd9_1600x1200.jpg', 'speaker': 'Erik Hersman', 'title': 'Reporting crisis via texting', 'duration': 236, 'slug': 'erik_hersman_on_reporting_crisis_via_texting', 'viewed_count': 403562}, {'id': 2, 'hero': ...",NaN,"['TED Fellows', 'design', 'global development', 'technology', 'water']",Meet the Water Canary,https://www.ted.com/talks/sonaar_luthra_meet_the_water_canary\n,353749
1192,122,"The Pirate Party fights for transparency, anonymity and sensible copyright laws. At TEDxObserver, Rick Falkvinge explains how he became the leader of Europe's tech-driven political party, which so far has won 17 seats across national parliaments in Europe.",1097,TEDxObserver,1331424000,10,Rick Falkvinge,Rick Falkvinge: I am a pirate,1,1333289675,"[{'id': 8, 'name': 'Informative', 'count': 156}, {'id': 7, 'name': 'Funny', 'count': 66}, {'id': 22, 'name': 'Fascinating', 'count': 54}, {'id': 3, 'name': 'Courageous', 'count': 148}, {'id': 10, 'name': 'Inspiring', 'count': 310}, {'id': 24, 'name': 'Persuasive', 'count': 185}, {'id': 9, 'name'...","[{'id': 1329, 'hero': 'https://pe.tedcdn.com/images/ted/b8b0b3e24229c2823cba77f2c3a00dc3b7c5ff70_800x600.jpg', 'speaker': 'Clay Shirky', 'title': 'Why SOPA is a bad idea', 'duration': 839, 'slug': 'defend_our_freedom_to_share_or_why_sopa_is_a_bad_idea', 'viewed_count': 1275141}, {'id': 871, 'her...",NaN,"['Internet', 'TEDx', 'global issues', 'politics', 'technology']",I am a pirate,https://www.ted.com/talks/rick_falkvinge_i_am_a_pirate\n,181010
1220,257,"As you surf the Web, information is being collected about you. Web tracking is not 100% evil -- personal data can make your browsing more efficient; cookies can help your favorite websites stay in business. But, says Gary Kovacs, it's your right to know what data is being collected about you. He...",399,TED2012,1330473600,32,Gary Kovacs,Gary Kovacs: Tracking our online trackers,1,1336057219,"[{'id': 23, 'name': 'Jaw-dropping', 'count': 910}, {'id': 22, 'name': 'Fascinating', 'count': 492}, {'id': 8, 'name': 'Informative', 'count': 1513}, {'id': 24, 'name': 'Persuasive', 'count': 745}, {'id': 3, 'name': 'Courageous', 'count': 256}, {'id': 9, 'name': 'Ingenious', 'count': 132}, {'id':...","[{'id': 1370, 'hero': 'https://pe.tedcdn.com/images/ted/db094e19db32cb842b63f52a39cf02858bf6457a_2880x1620.jpg', 'speaker': 'Avi Rubin', 'title': 'All your devices can be hacked', 'duration': 1016, 'slug': 'avi_rubin_all_your_devices_can_be_hacked', 'viewed_count': 1104692}, {'id': 788, 'hero': ...",NaN,"['Internet', 'advertising', 'business', 'privacy', 'technology']",Tracking our online trackers,https://www.ted.com/talks/gary_kovacs_tracking_the_trackers\n,2098639
1656,140,"In this lovely talk, TED Fellow Ryan Holladay shares his experiment with ""location-aware music."" This programming and musical feat involves hundreds of geotagged segments of sounds that only play when a listener is physically nearby, creating a magical sense of presence.",389,TED@BCG San Francisco,1383091200,33,Ryan Holladay,Ryan Holladay: To hear this

共有6条观察值缺失`speaker_occupation`变量值。

`speaker_occupation`表示演讲者的职业，不是挖掘最受欢迎的演讲主题的必要变量。并且从输出结果看，缺失`speaker_occupation`变量的观察值仍然是有效的。因此可以保留，无需进行处理。

#### 评估重复数据

根据数据变量的含义分析，虽然`description`、`name`、`title`、`url`是确定的唯一标识符，但是由于本数据集并不复合“每个单元格是一个值”这一数据整齐度标准，其中`ratings`、`related_talks`、`speaker_occupation`、`tags`四列对应各行观察值均包含多个变量，可拆分为多行，即`description`、`name`、`title`、`url`均有可重复。故在为拆分前，评估是否有重复数据。

In [37]:
ted_data[ted_data.duplicated(subset=["description", "name", "title", "url"])]

,comments,description,duration,event,film_date,languages,main_speaker,name,num_speaker,published_date,ratings,related_talks,speaker_occupation,tags,title,url,views


可见无重复数据，无需进行处理。

#### 评估不一致数据

本数据集中无不一致数据，无需评估。

#### 评估无效或错误数据

可以通过DataFrame的describe方法，对数值统计信息进行快速了解。

In [38]:
ted_data.describe()

,comments,duration,film_date,languages,num_speaker,published_date,views
count,2550.000000,2550.000000,2.550000e+03,2550.000000,2550.000000,2.550000e+03,2.550000e+03
mean,191.562353,826.510196,1.321928e+09,27.326275,1.028235,1.343525e+09,1.698297e+06
std,282.315223,374.009138,1.197391e+08,9.563452,0.207705,9.464009e+07,2.498479e+06
min,2.000000,135.000000,7.464960e+07,0.000000,1.000000,1.151367e+09,5.044300e+04
25%,63.000000,577.000000,1.257466e+09,23.000000,1.000000,1.268463e+09,7.557928e+05
50%,118.000000,848.000000,1.333238e+09,28.000000,1.000000,1.340935e+09,1.124524e+06
75%,221.750000,1046.750000,1.412964e+09,33.000000,1.000000,1.423432e+09,1.700760e+06
max,6404.000000,5256.000000,1.503792e+09,72.000000,5.000000,1.506092e+09,4.722711e+07


未见异常数据，无需处理。

## 清理数据

根据前面数据评估结论，我们需要进行的数据清理工作包括：

- 对`ratings`变量进行行拆分
- 对`related_talks`变量进行行拆分
- 对`speaker_occupation`变量进行行拆分
- 对`tags`变量进行行拆分
- 将`film_date`变量数据类型转换为日期时间
- 将`published_date`变量数据类型转换为日期时间

为了区分开经过清理后的数据和原始数据，我们创建新的变量`cleaned_ted_data`，让它为`ted_data`复制出的副本，我们之后的清理操作都将被运用到`cleaned_ted_data`上。

In [39]:
cleaned_ted_data = ted_data.copy()
cleaned_ted_data.head()

,comments,description,duration,event,film_date,languages,main_speaker,name,num_speaker,published_date,ratings,related_talks,speaker_occupation,tags,title,url,views
0,4553,Sir Ken Robinson makes an entertaining and profoundly moving case for creating an education system that nurtures (rather than undermines) creativity.,1164,TED2006,1140825600,60,Ken Robinson,Ken Robinson: Do schools kill creativity?,1,1151367060,"[{'id': 7, 'name': 'Funny', 'count': 19645}, {'id': 1, 'name': 'Beautiful', 'count': 4573}, {'id': 9, 'name': 'Ingenious', 'count': 6073}, {'id': 3, 'name': 'Courageous', 'count': 3253}, {'id': 11, 'name': 'Longwinded', 'count': 387}, {'id': 2, 'name': 'Confusing', 'count': 242}, {'id': 8, 'name...","[{'id': 865, 'hero': 'https://pe.tedcdn.com/images/ted/172559_800x600.jpg', 'speaker': 'Ken Robinson', 'title': 'Bring on the learning revolution!', 'duration': 1008, 'slug': 'sir_ken_robinson_bring_on_the_revolution', 'viewed_count': 7266103}, {'id': 1738, 'hero': 'https://pe.tedcdn.com/images/...",Author/educator,"['children', 'creativity', 'culture', 'dance', 'education', 'parenting', 'teaching']",Do schools kill creativity?,https://www.ted.com/talks/ken_robinson_says_schools_kill_creativity\n,47227110
1,265,"With the same humor and humanity he exuded in ""An Inconvenient Truth,"" Al Gore spells out 15 ways that individuals can address climate change immediately, from buying a hybrid to inventing a new, hotter brand name for global warming.",977,TED2006,1140825600,43,Al Gore,Al Gore: Averting the climate crisis,1,1151367060,"[{'id': 7, 'name': 'Funny', 'count': 544}, {'id': 3, 'name': 'Courageous', 'count': 139}, {'id': 2, 'name': 'Confusing', 'count': 62}, {'id': 1, 'name': 'Beautiful', 'count': 58}, {'id': 21, 'name': 'Unconvincing', 'count': 258}, {'id': 11, 'name': 'Longwinded', 'count': 113}, {'id': 8, 'name': ...","[{'id': 243, 'hero': 'https://pe.tedcdn.com/images/ted/566c14767bd62c5ff760e483c5b16cd2753328cd_2880x1620.jpg', 'speaker': 'Al Gore', 'title': 'New thinking on the climate crisis', 'duration': 1674, 'slug': 'al_gore_s_new_thinking_on_the_climate_crisis', 'viewed_count': 1751408}, {'id': 547, 'he...",Climate advocate,"['alternative energy', 'cars', 'climate change', 'culture', 'environment', 'global issues', 'science', 'sustainability', 'technology']",Averting the climate crisis,https://www.ted.com/talks/al_gore_on_averting_climate_crisis\n,3200520
2,124,"New York Times columnist David Pogue takes aim at technology’s worst interface-design offenders, and provides encouraging examples of products that get it right. To funny things up, he bursts into song.",1286,TED2006,1140739200,26,David Pogue,David Pogue: Simplicity sells,1,1151367060,"[{'id': 7, 'name': 'Funny', 'count': 964}, {'id': 3, 'name': 'Courageous', 'count': 45}, {'id': 9, 'name': 'Ingenious', 'count': 183}, {'id': 1, 'name': 'Beautiful', 'count': 60}, {'id': 21, 'name': 'Unconvincing', 'count': 104}, {'id': 11, 'name': 'Longwinded', 'count': 78}, {'id': 8, 'name': '...","[{'id': 1725, 'hero': 'https://pe.tedcdn.com/images/ted/b7f415a054cc0a2bfdd90d0ad5a7f64cf060150d_1600x1200.jpg', 'speaker': 'David Pogue', 'title': '10 top time-saving tech tips', 'duration': 344, 'slug': 'david_pogue_10_top_time_saving_tech_tips', 'viewed_count': 4843421}, {'id': 2274, 'hero': ...",Technology columnist,"['computers', 'entertainment', 'interface design', 'media', 'music', 'performance', 'simplicity', 'software', 'technology']",Simplicity sells,https://www.ted.com/talks/david_pogue_says_simplicity_sells\n,1636292
3,200,"In an emotionally charged talk, MacArthur-winning activist Majora Carter details her fight for environmental justice in the South Bronx -- and shows how minority neighborhoods suffer most from flawed urban policy.",1116,TED2006,1140912000,35,Majora Carter,Majora Carter: Greening the ghetto,1,1151367060,"[{'id': 3, 'name': 'Courageous', 'count': 760}, {'id': 1, 'name': 'Beautiful', 'count': 291}, {'id': 2, 'name': 'Confusing', 'count': 32}, {'i

对`ratings`、`related_talks`、`tags`变量进行行拆分：

`ratings`、`related_talks`、`tags`三个变量表面上看是列表，实际数据类型为字符串，需使用Python内置的`eval`函数，把字符串转换成表达式，可以帮我们把表示列表的字符串转换成列表本身。

In [40]:
cleaned_ted_data["ratings"] = cleaned_ted_data["ratings"].apply(lambda val: eval(val))

In [41]:
cleaned_ted_data["related_talks"] = cleaned_ted_data["related_talks"].apply(lambda val: eval(val))

In [42]:
cleaned_ted_data["tags"] = cleaned_ted_data["tags"].apply(lambda val: eval(val))

In [33]:
cleaned_ted_data = cleaned_ted_data.explode(["ratings"])

In [34]:
cleaned_ted_data = cleaned_ted_data.explode(["related_talks"])

In [35]:
cleaned_ted_data = cleaned_ted_data.explode(["tags"])

In [37]:
cleaned_ted_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1584296 entries, 0 to 2549
Data columns (total 17 columns):
 #   Column              Non-Null Count    Dtype 
---  ------              --------------    ----- 
 0   comments            1584296 non-null  int64 
 1   description         1584296 non-null  object
 2   duration            1584296 non-null  int64 
 3   event               1584296 non-null  object
 4   film_date           1584296 non-null  int64 
 5   languages           1584296 non-null  int64 
 6   main_speaker        1584296 non-null  object
 7   name                1584296 non-null  object
 8   num_speaker         1584296 non-null  int64 
 9   published_date      1584296 non-null  int64 
 10  ratings             1584296 non-null  object
 11  related_talks       1584296 non-null  object
 12  speaker_occupation  1581944 non-null  object
 13  tags                1584296 non-null  object
 14  title               1584296 non-null  object
 15  url                 1584296 non-null  ob

对`speaker_occupation`变量进行行拆分：

检查原始数据集中`speaker_occupation`不同职业出现的次数。

In [39]:
ted_data["speaker_occupation"].value_counts()

speaker_occupation
Writer                           45
Designer                         34
Artist                           34
Journalist                       33
Entrepreneur                     31
                                 ..
Urban designer                    1
Mayor of Albuquerque              1
Developmental psychologist        1
Student, musician                 1
Mathematician, data scientist     1
Name: count, Length: 1458, dtype: int64

可见，`speaker_occupation`共有1458种，通过随机抽取10行，观察`speaker_occupation`变量每个单元格内不同职业的分隔方式。

In [85]:
ted_data["speaker_occupation"].sample(10)

844                 Nobel peace laureate
954                               Artist
2324                        Psychologist
331                        Planet walker
8       Philosopher, cognitive scientist
1456                  Musician, activist
831                  Website co-founders
2462                Grandmaster, analyst
1150      Director of The British Museum
546                        Mathematician
Name: speaker_occupation, dtype: object

经多组观察，每个单元格内不同职业可以","、"/"、"and"、"+"进行分隔。分隔后对每个职业的首字母大写。

In [86]:
import re
# 定义处理函数：拆分职业并统一首字母大写
def process_occupation(occupation):
    if pd.isna(occupation):
        return []

    # 正则表达式增加对+的处理，现在包含：,、/、and、+四种分隔符
    split_occupations = re.split(r'[,/+]\s*|\s+and\s+', str(occupation))

    # 处理每个职业：去除空格并首字母大写
    processed = []
    for occ in split_occupations:
        occ = occ.strip()  # 去除前后空格
        if occ:  # 确保不是空字符串
            occ = occ.capitalize()  # 首字母大写，其余小写
            processed.append(occ)

    return processed

In [87]:
cleaned_ted_data["speaker_occupation"] = cleaned_ted_data["speaker_occupation"].apply(process_occupation)

In [100]:
cleaned_ted_data = cleaned_ted_data.explode("speaker_occupation")

In [111]:
cleaned_ted_data["speaker_occupation"].sample(10)

2155    Principal investigator
1247                  Designer
1652         Language advocate
1789                    Artist
2127                Roboticist
923          Software engineer
1222                   Blogger
1585    False memories scholar
223                        Ceo
537           Sound consultant
Name: speaker_occupation, dtype: object

经多次观察，符合预期。

将`film_date`变量数据类型转换为日期时间

In [112]:
cleaned_ted_data["film_date"] = pd.to_datetime(cleaned_ted_data["film_date"])
cleaned_ted_data["film_date"]

0      1970-01-01 00:00:01.140825600
0      1970-01-01 00:00:01.140825600
0      1970-01-01 00:00:01.140825600
0      1970-01-01 00:00:01.140825600
0      1970-01-01 00:00:01.140825600
                    ...             
2549   1970-01-01 00:00:01.492992000
2549   1970-01-01 00:00:01.492992000
2549   1970-01-01 00:00:01.492992000
2549   1970-01-01 00:00:01.492992000
2549   1970-01-01 00:00:01.492992000
Name: film_date, Length: 2019668, dtype: datetime64[ns]

将`published_date`变量数据类型转换为日期时间

In [114]:
cleaned_ted_data["published_date"] = pd.to_datetime(cleaned_ted_data["published_date"])
cleaned_ted_data["published_date"]

0      1970-01-01 00:00:01.151367060
0      1970-01-01 00:00:01.151367060
0      1970-01-01 00:00:01.151367060
0      1970-01-01 00:00:01.151367060
0      1970-01-01 00:00:01.151367060
                    ...             
2549   1970-01-01 00:00:01.506092422
2549   1970-01-01 00:00:01.506092422
2549   1970-01-01 00:00:01.506092422
2549   1970-01-01 00:00:01.506092422
2549   1970-01-01 00:00:01.506092422
Name: published_date, Length: 2019668, dtype: datetime64[ns]

## 保存清理后的数据

完成数据清理后，把干净整齐的数据保存在新的文件里，文件名为`cleaned_ted.csv`。

In [115]:
cleaned_ted_data.head()

,comments,description,duration,event,film_date,languages,main_speaker,name,num_speaker,published_date,ratings,related_talks,speaker_occupation,tags,title,url,views
0,4553,Sir Ken Robinson makes an entertaining and profoundly moving case for creating an education system that nurtures (rather than undermines) creativity.,1164,TED2006,1970-01-01 00:00:01.140825600,60,Ken Robinson,Ken Robinson: Do schools kill creativity?,1,1970-01-01 00:00:01.151367060,"{'id': 7, 'name': 'Funny', 'count': 19645}","{'id': 865, 'hero': 'https://pe.tedcdn.com/images/ted/172559_800x600.jpg', 'speaker': 'Ken Robinson', 'title': 'Bring on the learning revolution!', 'duration': 1008, 'slug': 'sir_ken_robinson_bring_on_the_revolution', 'viewed_count': 7266103}",Author,children,Do schools kill creativity?,https://www.ted.com/talks/ken_robinson_says_schools_kill_creativity\n,47227110
0,4553,Sir Ken Robinson makes an entertaining and profoundly moving case for creating an education system that nurtures (rather than undermines) creativity.,1164,TED2006,1970-01-01 00:00:01.140825600,60,Ken Robinson,Ken Robinson: Do schools kill creativity?,1,1970-01-01 00:00:01.151367060,"{'id': 7, 'name': 'Funny', 'count': 19645}","{'id': 865, 'hero': 'https://pe.tedcdn.com/images/ted/172559_800x600.jpg', 'speaker': 'Ken Robinson', 'title': 'Bring on the learning revolution!', 'duration': 1008, 'slug': 'sir_ken_robinson_bring_on_the_revolution', 'viewed_count': 7266103}",Educator,children,Do schools kill creativity?,https://www.ted.com/talks/ken_robinson_says_schools_kill_creativity\n,47227110
0,4553,Sir Ken Robinson makes an entertaining and profoundly moving case for creating an education system that nurtures (rather than undermines) creativity.,1164,TED2006,1970-01-01 00:00:01.140825600,60,Ken Robinson,Ken Robinson: Do schools kill creativity?,1,1970-01-01 00:00:01.151367060,"{'id': 7, 'name': 'Funny', 'count': 19645}","{'id': 865, 'hero': 'https://pe.tedcdn.com/images/ted/172559_800x600.jpg', 'speaker': 'Ken Robinson', 'title': 'Bring on the learning revolution!', 'duration': 1008, 'slug': 'sir_ken_robinson_bring_on_the_revolution', 'viewed_count': 7266103}",Author,creativity,Do schools kill creativity?,https://www.ted.com/talks/ken_robinson_says_schools_kill_creativity\n,47227110
0,4553,Sir Ken Robinson makes an entertaining and profoundly moving case for creating an education system that nurtures (rather than undermines) creativity.,1164,TED2006,1970-01-01 00:00:01.140825600,60,Ken Robinson,Ken Robinson: Do schools kill creativity?,1,1970-01-01 00:00:01.151367060,"{'id': 7, 'name': 'Funny', 'count': 19645}","{'id': 865, 'hero': 'https://pe.tedcdn.com/images/ted/172559_800x600.jpg', 'speaker': 'Ken Robinson', 'title': 'Bring on the learning revolution!', 'duration': 1008, 'slug': 'sir_ken_robinson_bring_on_the_revolution', 'viewed_count': 7266103}",Educator,creativity,Do schools kill creativity?,https://www.ted.com/talks/ken_robinson_says_schools_kill_creativity\n,47227110
0,4553,Sir Ken Robinson makes an entertaining and profoundly moving case for creating an education system that nurtures (rather than undermines) creativity.,1164,TED2006,1970-01-01 00:00:01.140825600,60,Ken Robinson,Ken Robinson: Do schools kill creativity?,1,1970-01-01 00:00:01.151367060,"{'id': 7, 'name': 'Funny', 'count': 19645}","{'id': 865, 'hero': 'https://pe.tedcdn.com/images/ted/172559_800x600.jpg', 'speaker': 'Ken Robinson', 'title': 'Bring on the learning revolution!', 'duration': 1008, 'slug': 'sir_ken_robinson_bring_on_the_revolution', 'viewed_count': 7266103}",Author,culture,Do schools kill creativity?,https://www.ted.com/talks/ken_robinson_says_schools_kill_creativity\n,47227110


In [116]:
cleaned_ted_data.to_csv("cleaned_ted.csv", index=False)

In [117]:
pd.read_csv("cleaned_ted.csv").head()

,comments,description,duration,event,film_date,languages,main_speaker,name,num_speaker,published_date,ratings,related_talks,speaker_occupation,tags,title,url,views
0,4553,Sir Ken Robinson makes an entertaining and profoundly moving case for creating an education system that nurtures (rather than undermines) creativity.,1164,TED2006,1970-01-01 00:00:01.140825600,60,Ken Robinson,Ken Robinson: Do schools kill creativity?,1,1970-01-01 00:00:01.151367060,"{'id': 7, 'name': 'Funny', 'count': 19645}","{'id': 865, 'hero': 'https://pe.tedcdn.com/images/ted/172559_800x600.jpg', 'speaker': 'Ken Robinson', 'title': 'Bring on the learning revolution!', 'duration': 1008, 'slug': 'sir_ken_robinson_bring_on_the_revolution', 'viewed_count': 7266103}",Author,children,Do schools kill creativity?,https://www.ted.com/talks/ken_robinson_says_schools_kill_creativity\n,47227110
1,4553,Sir Ken Robinson makes an entertaining and profoundly moving case for creating an education system that nurtures (rather than undermines) creativity.,1164,TED2006,1970-01-01 00:00:01.140825600,60,Ken Robinson,Ken Robinson: Do schools kill creativity?,1,1970-01-01 00:00:01.151367060,"{'id': 7, 'name': 'Funny', 'count': 19645}","{'id': 865, 'hero': 'https://pe.tedcdn.com/images/ted/172559_800x600.jpg', 'speaker': 'Ken Robinson', 'title': 'Bring on the learning revolution!', 'duration': 1008, 'slug': 'sir_ken_robinson_bring_on_the_revolution', 'viewed_count': 7266103}",Educator,children,Do schools kill creativity?,https://www.ted.com/talks/ken_robinson_says_schools_kill_creativity\n,47227110
2,4553,Sir Ken Robinson makes an entertaining and profoundly moving case for creating an education system that nurtures (rather than undermines) creativity.,1164,TED2006,1970-01-01 00:00:01.140825600,60,Ken Robinson,Ken Robinson: Do schools kill creativity?,1,1970-01-01 00:00:01.151367060,"{'id': 7, 'name': 'Funny', 'count': 19645}","{'id': 865, 'hero': 'https://pe.tedcdn.com/images/ted/172559_800x600.jpg', 'speaker': 'Ken Robinson', 'title': 'Bring on the learning revolution!', 'duration': 1008, 'slug': 'sir_ken_robinson_bring_on_the_revolution', 'viewed_count': 7266103}",Author,creativity,Do schools kill creativity?,https://www.ted.com/talks/ken_robinson_says_schools_kill_creativity\n,47227110
3,4553,Sir Ken Robinson makes an entertaining and profoundly moving case for creating an education system that nurtures (rather than undermines) creativity.,1164,TED2006,1970-01-01 00:00:01.140825600,60,Ken Robinson,Ken Robinson: Do schools kill creativity?,1,1970-01-01 00:00:01.151367060,"{'id': 7, 'name': 'Funny', 'count': 19645}","{'id': 865, 'hero': 'https://pe.tedcdn.com/images/ted/172559_800x600.jpg', 'speaker': 'Ken Robinson', 'title': 'Bring on the learning revolution!', 'duration': 1008, 'slug': 'sir_ken_robinson_bring_on_the_revolution', 'viewed_count': 7266103}",Educator,creativity,Do schools kill creativity?,https://www.ted.com/talks/ken_robinson_says_schools_kill_creativity\n,47227110
4,4553,Sir Ken Robinson makes an entertaining and profoundly moving case for creating an education system that nurtures (rather than undermines) creativity.,1164,TED2006,1970-01-01 00:00:01.140825600,60,Ken Robinson,Ken Robinson: Do schools kill creativity?,1,1970-01-01 00:00:01.151367060,"{'id': 7, 'name': 'Funny', 'count': 19645}","{'id': 865, 'hero': 'https://pe.tedcdn.com/images/ted/172559_800x600.jpg', 'speaker': 'Ken Robinson', 'title': 'Bring on the learning revolution!', 'duration': 1008, 'slug': 'sir_ken_robinson_bring_on_the_revolution', 'viewed_count': 7266103}",Author,culture,Do schools kill creativity?,https://www.ted.com/talks/ken_robinson_says_schools_kill_creativity\n,47227110


数据评估与清理工作完毕。